perf(async): Introduce async and timeout support

alexmaco opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
Currently, this will:
* replace rayon with async_std
* adds support for specifying a per-module "timeout" key in config
* adds support for specifying a global "prompt_timeout" key in config, used for all modules that don't have a "timeout" explicitly given
* changes all modules to execute subprocesses and file reads asynchronously, and library calls on a different future, so the timeouts work
* report an exceeded timeout value when running `timings`

A quick measurement shows no visible performance impact https://github.com/starship/starship/issues/312#issuecomment-767142122

Feedback welcome as to exactly how to deal with timeouts, and when they should be silent or not. The current version is geared for debugging, so it's immediately obvious which module is causing the prompt to hang.

#### Motivation and Context
Async allows pervasive and simple handling of timeouts for virtually any code. There are multiple open issues related to responsiveness, and this approach with allow timeout control with arbitrary granularity, while maintaining parallelism.

<!--- If it fixes an open issue, please link to the issue here. -->
Closes #312, #597, #891, #940, #1551
Will possibly close a few others once fully implemented.

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

davidkna commented:
Thanks for your work on this PR so far. I think you can work on completing this. But can you elaborate what you mean regarding "refinements for git_status"?

alexmaco commented:
> Thanks for your work on this PR so far. I think you can work on completing this. 

Thanks! I'll switch a few more modules to async, so we can have a relatively complete set of cases covered as a starting point.

> But can you elaborate what you mean regarding "refinements for git_status"?

I guess that comment is outdated. I think all that's left to add is the "timeout" variable in the docs for the modules where it now works (git_status, rust, elixir).

If you have any remarks relative to strategies or individual cases, I look forward to reading them.

davidkna commented on src/formatter/string_formatter.rs:
```diff
@@ -1,6 +1,7 @@
+use crate::utils::parallel_map;
 use ansi_term::Style;
+use core::future::Future;
```
Should this be `std`?
```suggestion
use std::future::Future;
```

davidkna commented on src/modules/elixir.rs:
```diff
@@ -4,9 +4,15 @@ use crate::configs::elixir::ElixirConfig;
 use regex::Regex;
-use std::ops::Deref;
+
+lazy_static! {
+    static ref VERSIONS: AsyncOnce<Option<(String, String)>> =
+        AsyncOnce::new(async { get_elixir_version().await });
+}
```
Doesn't `async` already have semantics that would make a suitable as a replacement to `Lazy` here? IIRC `async` functions don't get executed until they get polled. Or does `.await` imply taking ownership?

davidkna reviewed (commented):
Just had a quick look at your work

davidkna commented on src/modules/mod.rs:
```diff
@@ -130,6 +133,32 @@ pub fn handle<'a>(module: &str, context: &'a Context) -> Option<Module<'a>> {
 
+    let (time_limit, user_specified) = match context.module_timeout(module) {
+        Some(dur) => (dur, true),
+        None => (Duration::from_millis(3000), false),
```
I'd prefer a default timeout closer to the current command timeout of 500ms.

alexmaco replied:
I've set it to 700, lmk if that sounds ok.

alexmaco commented:
@davidkna I've moved some more modules to async, and updated the PR description to reflect it's no longer WIP.

Let me know what other issues you spot, or if you think some more timeout-specific documentation should be added.

alexmaco commented on src/modules/elixir.rs:
```diff
@@ -4,9 +4,14 @@ use crate::configs::elixir::ElixirConfig;
 use regex::Regex;
-use std::ops::Deref;
+
+lazy_static! {
+    static ref VERSIONS: AsyncOnce<Option<(String, String)>> = AsyncOnce::new(get_elixir_version());
+}
```
I've just seen that utils/Context PR merge, indeed, it's going to need some reworking.

So you're saying it's ok to depend on `futures`, in addition to `async_std`, correct? I had actually tried to avoid adding it in the first pass :smile:

davidkna replied:
I adding it should be fine, but I don't have much experience with async in rust. I didn't really notice much of an impact on compile times, is there any other reason to avoid it?

alexmaco replied:
I'm also not seeing increased compile times. The idea was just to keep dependencies focused, but otherwise there shouldn't be any issues. It will actually improve some areas.

alexmaco commented on src/logger.rs:
```diff
@@ -109,8 +109,8 @@ impl log::Log for StarshipLogger {
                 match record.level() {
-                    Level::Trace => Color::Black.dimmed().paint(format!("{}", record.level())),
-                    Level::Debug => Color::Black.paint(format!("{}", record.level())),
+                    Level::Trace => Color::Blue.dimmed().paint(format!("{}", record.level())),
+                    Level::Debug => Color::Cyan.paint(format!("{}", record.level())),
```
Moved to https://github.com/starship/starship/pull/2299

matchai commented:
> - adds support for specifying a per-module "timeout" key in config

What would be the value in having per-module timeouts rather than prompt-wide timeouts? In what circumstances would someone want different timeouts on different modules? 

I'm not sure I see the value in that distinction, while per-module timeouts increase config and doc complexity.

alexmaco commented:
> What would be the value in having per-module timeouts rather than prompt-wide timeouts? In what circumstances would someone want different timeouts on different modules?

Per-module timeouts was just my starting assumption. I figured I would set smaller timeouts to modules that fire in dirs I used frequently, but it's ok for others, in less frequently used dirs to take longer. I'm now using such a config, and it does work. Otoh, a global timeout does sound like the first thing a new user would reach for if they get annoyed with the response time.

> I'm not sure I see the value in that distinction, while introducing per-module timeouts increases config and doc complexity.

I should have probably raised this point earlier. Imho, the docs should specify keys like `disabled` or `timeout` in a general way, for all modules, owing primarily to the fact that descriptions are very repetitive. I'm not sure this is the place for discussing `disabled` however.

In any case, a global `prompt_timeout` or just `timeout` key seems very pertinent, I'll add it.

As for per-module timeout, I at least find it useful, and the implementation cost is only a per-module config lookup, so not above 10 lines. Is docs verbosity the main concern here? If so, how about having per-module `timeout` explained in the `Prompt` section or above?

alexmaco commented:
If anyone's wondering about all the `async_*` function prefixes, they are not idiomatic. Ideally there should only be one `async fn exec_cmd(...)` and not a `fn exec_cmd(...)` and an `async fn async_exec_cmd(...)`.

I've only left both in to avoid making the PR gigantic, and enable piecemeal conversion of modules. Eventually, when everything is asyncified, the `async_` prefix should be dropped with a rename.

alexmaco commented:
@matchai 

Added a global `prompt_timeout`, effectively making the previous 700ms default fallback configurable.
Also, timeout overruns are now shown with the `timings` command, and `timings` and `explain` do not obey timeout restrictions.

With `prompt_timeout = 50`, it looks like this:

![image](https://user-images.githubusercontent.com/24556329/107815726-6f8fcf00-6d7c-11eb-8aeb-d0fab4c38c7e.png)

davidkna commented on src/modules/rust.rs:
```diff
@@ -100,10 +114,11 @@ fn env_rustup_toolchain(context: &Context) -> Option<String> {
 
-fn execute_rustup_override_list(cwd: &Path) -> Option<String> {
+async fn execute_rustup_override_list(cwd: &Path) -> Option<String> {
     let Output { stdout, .. } = Command::new("rustup")
```
I think `exec_cmd` already handles errors sufficiently by logging them. If you think the real error is required here, I'd prefer you add a second method to `Context`.
Beyond that: I think for modules that are async it makes sense to ignore `command_timeout` because having multiple separate timeouts for a single module would get confusing.

alexmaco commented:
@davidkna Afaict the only outstanding issue right now is the discussion related to per-module timeouts, pointed out by @matchai .
What I think we should decide is whether to mention `timeout` once at the beginning, or to list it for every module. Alternatively, it's also possible to leave it undocumented, or remove it if you really don't want it, but I suspect power users will want it, especially for custom modules.

There are several more modules to switch to async, and some post-asyncification cleanup, but I'm not sure I should include _everything_ in this one PR.

LMK if you have any more thoughts.

matchai commented:
>  Alternatively, it's also possible to leave it undocumented, or remove it if you really don't want it, but I suspect power users will want it, especially for custom modules.

I would suggest only documenting it in the `custom` module, and leaving it undocumented in the others. What do others think?

alexmaco commented:
> I would suggest only documenting it in the `custom` module, and leaving it undocumented in the others. What do others think?

I think that sounds reasonable, docs updated.

alexmaco commented:
@davidkna Unless I've missed anything, asyncification should now be complete.

matchai commented:
@starship/astronauts Since this is such a hefty refactor, could contributors please try installing this fork locally and share benchmarks across various systems and configurations?

I'm personally seeing minor performance degradations, but nothing too severe.
Running on a Macbook Pro 2018, 2.2Ghz, macOS 11.2.1:
```
starship on  timeout is  v0.50.0 via  v1.50.0
❯ hyperfine "starship prompt" -m 100
Benchmark #1: starship prompt
  Time (mean ± σ):     127.6 ms ±   3.3 ms    [User: 111.9 ms, System: 34.4 ms]
  Range (min … max):   121.5 ms … 139.5 ms    100 runs

starship on  master is  v0.50.0 via  v1.50.0
❯ hyperfine "starship prompt" -m 100
Benchmark #1: starship prompt
  Time (mean ± σ):     124.6 ms ±   4.2 ms    [User: 110.0 ms, System: 33.5 ms]
  Range (min … max):   117.9 ms … 141.4 ms    100 runs
```

davidkna commented on src/utils.rs:
```diff
@@ -331,13 +371,47 @@ fn internal_exec_cmd(cmd: &str, args: &[&str], time_limit: Duration) -> Option<C
+    use async_std::task::block_on;
+
     use super::*;
+    fn internal_exec_cmd(cmd: &str, args: &[&str]) -> Option<CommandOutput> {
```
I think it would be better to make the tests use [`async_std::test`](https://docs.rs/async-std/1.9.0/async_std/attr.test.html) here.

davidkna commented:
Tests with default config in the starship repo:

On an Intel i3-9100 with Arch Linux:
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `./starship-git prompt` | 73.9 ± 1.1 | 72.8 | 79.1 | 1.00 |
| `./starship-async prompt` | 76.8 ± 1.3 | 75.3 | 81.1 | 1.04 ± 0.02 |

On an Apple M1 with MacOS 11.2:
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `./starship-git prompt` | 61.9 ± 1.2 | 60.4 | 65.5 | 1.01 ± 0.03 |
| `./starship-async prompt` | 61.5 ± 1.3 | 60.1 | 65.7 | 1.00 |

davidkna commented on src/utils.rs:
```diff
@@ -331,13 +371,47 @@ fn internal_exec_cmd(cmd: &str, args: &[&str], time_limit: Duration) -> Option<C
+// TODO: replace this with parallel-stream (or similar) when possible.
+// At the time of writing parallel-stream 2.1.2 only works with an old
+// version of async_std, and is missing several features.
+pub async fn parallel_map<I, M, T, Fut, U, C>(items: I, map: M, mut collector: C)
```
I think you should use `futures::prelude::stream::Buffered` instead for this.

alexmaco commented on src/modules/mod.rs:
```diff
@@ -57,79 +57,104 @@ mod zig;
+            Ok(res) => res,
+            Err(TimeoutError { .. }) => {
+                log::warn!("Executing module {:?} timed out.", module);
+                log::warn!("You can set prompt_timeout in your config to a higher value to allow longer-running modules to keep executing.");
```
Oops, looks like I also missed something more in the custom module handling. Adding it.

alexmaco replied:
I've move this away into Context to enforce timeouts for custom modules. I had overlooked the fact that they didn't actually use the same path so no timeout applied to them. Should be fixed now.

andytom commented:
Running Fedora Linux AMD Ryzen 7 PRO 3700U w/ Radeon Vega Mobile Gfx

Tests with default config in the starship repo:
```
❯ hyperfine "./starship_master prompt" "./starship_2223 prompt"
Benchmark #1: ./starship_master prompt
  Time (mean ± σ):     115.2 ms ±  14.6 ms    [User: 102.4 ms, System: 26.3 ms]
  Range (min … max):    94.8 ms … 145.0 ms    26 runs
 
Benchmark #2: ./starship_2223 prompt
  Time (mean ± σ):     107.0 ms ±  10.9 ms    [User: 93.0 ms, System: 28.1 ms]
  Range (min … max):    95.1 ms … 135.5 ms    28 runs
 
Summary
  './starship_2223 prompt' ran
    1.08 ± 0.17 times faster than './starship_master prompt'
```

alexmaco commented:
I'm also seeing very small differences now with everything converted, but they seem to vary a lot. At worst it's something like:
```bash
❯ hyperfine -w 3 "./starship.master prompt" "./starship.timeout prompt"
Benchmark #1: ./starship.master prompt
  Time (mean ± σ):      95.7 ms ±   2.8 ms    [User: 86.0 ms, System: 20.0 ms]
  Range (min … max):    89.8 ms … 100.6 ms    29 runs
 
Benchmark #2: ./starship.timeout prompt
  Time (mean ± σ):      98.9 ms ±   5.8 ms    [User: 89.5 ms, System: 19.9 ms]
  Range (min … max):    90.9 ms … 114.5 ms    31 runs
 
Summary
  './starship.master prompt' ran
    1.03 ± 0.07 times faster than './starship.timeout prompt'
```

Profiling doesn't show anything egregious at first glance, aside from `git_status` being heavy, and the `OnceCell` in there possibly causing a little contention (usage of `OnceCell` predates async).

@davidkna I believe I haven't missed addressing any review comment. The only new thing really is the last commit that applies timeout to custom modules. While doing that I've tried to also run the custom modules in parallel, but the lifetime juggling is a little fierce in there at the moment, so I've left it out right now. The only downside is that timeouts for custom modules are effectively additive, which may be a little surprising until we fix it.

Outside of more review comments + fixes, I think the main question is whether:
1. this can be merged as such while opening a new issue for performance profiling and tweaks, or
1. the perf tweaks need to be part of this PR

Personally I'd support the former, since it would allow a more focused discussion and gathering (config,platform,timings) data sets.

chipbuster commented:
I think my opinion on this is that we don't have too much to worry about in terms of performance on UNIX-likes (I *would* like to see a Windows benchmark before deciding for sure). I'm not seeing much in the way of definite performance losses, so the question in my mind is if this is going to make it harder to add new features or modify the shell as it currently exists (due to lifetime issues, async complexity, etc.)

* * *

On a machine with an E3 Xeon and all directories accessed by NFS, starship repository:

```
> hyperfine -w 3 "../../starship-test/async/target/release/starship prompt" "../../starship-test/default/target/release/starship prompt"

Benchmark #1: ../../starship-test/async/target/release/starship prompt
  Time (mean ± σ):      76.7 ms ±   1.8 ms    [User: 63.4 ms, System: 24.6 ms]
  Range (min … max):    74.8 ms …  82.5 ms    39 runs
 
Benchmark #2: ../../starship-test/default/target/release/starship prompt
  Time (mean ± σ):      76.4 ms ±   2.5 ms    [User: 68.8 ms, System: 19.0 ms]
  Range (min … max):    74.0 ms …  86.2 ms    38 runs
 
Summary
  '../../starship-test/default/target/release/starship prompt' ran
    1.00 ± 0.04 times faster than '../../starship-test/async/target/release/starship prompt'
```

alexmaco commented:
> so the question in my mind is if this is going to make it harder to add new features or modify the shell as it currently exists (due to lifetime issues, async complexity, etc.)

I think this is definitely a valid concern. Working on this I've ran into error messages with significantly less polish that the non-async part of the language (some even talk about generators, even though that's just internal plumbing). Lifetime juggling, coupled with `Pin` usage has also been a practical learning experience, hence some extra cloning for simplification (i.e. `async_map` passing a `String`, even if `&str` is technically possible but more complicated).

That said, any suggestions on how to simplify or compartmentalize accidental async weirdness are welcome, I'd certainly like for it to be simpler than it is now.

alexmaco commented:
@matchai @davidkna 

Any other takes on what are the next steps to make progress with this PR?

So far, we have:
1. some performance degradation in the low percentages (visible only on some platforms, I think there is room for optimization, I have a few pending changes)
1. some concerns about added complexity by @chipbuster (valid IMO, though exactly what's suboptimal or how to address that is not immediately obvious; otoh, as async usage is becoming more common, perhaps this will become less and less of an issue?)
1. some tech debt on running `custom` modules in parallel (they still run sequentially)
1. the 700ms timeout value (I just ballparked something that worked for me, close to the previous 500ms per command)

chipbuster commented:
Hey @alexmaco,

I did some talking with @matchai yesterday. We don't think the perf degradation looks significant or that the custom modules are going to be a huge issue. It's mostly looking like complexity is going to be our biggest blocker here.

If we merge this, it's pretty quickly going to get away from a simple revert if we decide that the added complexity isn't desirable or sustainable. With that in mind: I have a question for you: do you have an idea for how hard this would be to do a piecewise reversion of?

Say that we merge this and find that the complexity is too much, but we've already heavily modified 10 modules. Is it possible to revert the old modules and then wrap the new ones with some sort of blocking executor and then feed all of it back into rayon?

If so, I'd be much less antsy about merging this, since if the challenges prove insurmountable, we can just convert the modules back to synchronous code at our own pace (instead of having to freeze the project, convert all the changes back into synchronous code, and merge them along with a rayon revert in one big PR).

alexmaco commented:
@chipbuster 

(forward disclaimer: I'll rebase this if you give the go-ahead, because it spams the conversation)

After sleeping on it, this is what I have so far:

### Reverting

Completely reverting async equates to porting the changes that happened in the meantime over the pre-merge sync code.

Reverting individual modules is relatively simple, modulo adjusting some function names, and wrapping in an executor.
Calling an `async` function from sync code only requires wrapping in `async_std::task::block_on()`.
If sync code needs to return a `Future` to some async code, that works trivially by wrapping a value in `std::future::ready`.

So, most of the complexity that may warrant reverting is in the the infrastructure.

Imho, most of the complexity is at top level (i.e. `print.rs` and `StringFormatter`), rather than the functionality leaves (e.g. `utils.rs` and `Context`).
Reverting either the toplevel or toplevel+utils would be a choice between:
1. moving the timeout functionality down into `Context`/`utils.rs` (which would be kept async, and only the top level would be rayon)
2. wrapping up the incoming/outgoing interfaces to the module code, and reverting both toplevel and utils to sync code (backporting any eventual changes)

The first would require less sync-async api wrapping for the modules, while the second would require less dealing with async overall (module wrapping is simple and repetitive).

### Future simplification

Another approach that may simplify things is reworking the `StringFormatter` api.
This would also work hand-in-hand with changing the module activation (i.e. let modules define conditions they are sensitive to, and only checking for those instead of reading all files from the current directory).
This would imply a 2-stage nature for modules, a per-module `Requirements`, managed by the replacement of StringFormatter that also does lazy dir scanning, and `Module` proper, which is fired lazily, without the need for `parallel_map`, which infects some code with undue `'static` requirements.

However I don't yet have a prototype here, I was waiting on async, but I do think this would get rid of the complexity at top level.
Goes without saying that it would be a significant rework in itself.

### Misc performance

While measuring another project, I've come to suspect `tokio` may actually be faster than `async_std`.
The api however would be slightly more complicated (more different from `std`, requires keeping a `Runtime` on `Context`, etc.).
So probably not the first choice since the blocker is complexity, but I thought to mention it.

betaboon commented:
I'm very much looking forward to seeing this merged at some point. this is the only feature in starship that I'm missing which prevents me from migrating to starship ;)

keep up the good work

Yash-Singh1 commented:
Any updates on this PR?

matchai commented:
To clear up any misunderstanding, this PR doesn't implement an async prompt, but introduces async Rust code.

Unfortunately, this PR has since gotten very stale, having been made over a year ago. We now have command timeouts, which address most of the purpose that timeouts would serve in this PR.

Yash-Singh1 commented:
So is there any plans to support async loading of ZSH plugins?

matchai commented:
There are no immediate plans to add async support.
At the end of the day, async is a means to an end – a way of achieving a fast prompt. As much as I'd love to add support for async, we haven't found a good way to achieve it in a way that is cross-shell.

For 99% of users, Starship is instant. We're continuing to work on performance improvements that would eliminate the need for async.
